crates/node/src/networking/download_manager.rs

```diff
@@ -8,12 +9,101 @@ use hyper::server::conn::http1;
+    file: &str,
+    http_peer_list: Vec<(SocketAddr, Option<u16>)>,
+    http_client: &reqwest::Client,
+    file_hash: gevulot_node::types::Hash,
+) -> Result<()> {
+    tracing::trace!("download_file url:{url} local_directory_path:{local_directory_path:?} file:{file} file_hash:{file_hash} http_peer_list:{http_peer_list:?}");
+    let url = reqwest::Url::parse(url)?;
+    let mut resp = match http_client.get(url.clone()).send().await {
+        Ok(resp) => resp,
+        Err(err) => {
+            let peer_urls: Vec<reqwest::Url> = http_peer_list
+                .iter()
+                .filter_map(|(peer, port)| {
+                    port.map(|port| {
+                        //use parse to create an URL, no new method.
+                        let mut url = reqwest::Url::parse("http://localhost").unwrap(); //unwrap always succeed
+                        url.set_ip_host(peer.ip()).unwrap(); //unwrap always succeed
+                        url.set_port(Some(port)).unwrap(); //unwrap always succeed
+                        url.set_path(file); //unwrap always succeed
+                        url
+                    })
+                })
+                .collect();
+            let mut resp = None;
+            for url in peer_urls {
+                if let Ok(val) = http_client.get(url.clone()).send().await {
+                    resp = Some(val);
+                    break;
+                }
+            }
+            match resp {
+                Some(resp) => resp,
+                _ => {
+                    return Err(eyre!(
+                        "Download no host found to download the file: {:?}",
+                        file
+                    ));
+                }
+            }
+        }
+    };
+
+    if resp.status() == reqwest::StatusCode::OK {
+        let file_path = local_directory_path.join(file);
+        // Ensure any necessary subdirectories exists.
+        if let Some(parent) = file_path.parent() {
+            if let Ok(false) = tokio::fs::try_exists(parent).await {
+                tokio::fs::create_dir_all(parent).await?;
+            }
+        }
+        //create a tmp file during download.
+        //this way the file won't be available for download from the other nodes
+        //until it is completely written.
+        let mut tmp_file_path = file_path.clone();
+        tmp_file_path.set_extension("tmp");
+        let fd = tokio::fs::File::create(&tmp_file_path).await?;
+        let mut fd = tokio::io::BufWriter::new(fd);
+
+        //create the Hasher to verify the Hash
+        let mut hasher = blake3::Hasher::new();
+
+        while let Some(chunk) = resp.chunk().await? {
+            hasher.update(&chunk);
+            fd.write_all(&chunk).await?;
+        }
+
+        fd.flush().await?;
+        let checksum: gevulot_node::types::Hash = (&hasher.finalize()).into();
+        if checksum != file_hash {
+            Err(eyre!("Download file: {:?}, bad checksum", file))
```

Comment:
The downloaded file is left on disk even if it failed the checksum. Should it be deleted or should just account for it in some kind of garbage collection if we implement such in the future?

Reply:
Yes, the file should be removed, and the program removed or inactivate too because currently, if the error arrives that download will always restart because the program is written in the db and the scheduler try to restart it.
I remove the program too or inactivate some way the program.